sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/AuthClient.java

```diff
@@ -0,0 +1,182 @@
+
+    /**
+     * Constructor.
+     *
+     * <p>
+     * The constructor creates a default RestClient.
+     * </p>
+     */
+    AuthClient() {
+        super(RestClientFactory.createClient());
+    }
+
+    /**
+     * Get an access token for a managed identity.
+     *
+     * @param resource the resource.
+     * @return the authorization token.
+     */
+    public String getAccessToken(String resource) {
+        String result;
+
+        if (System.getenv("WEBSITE_SITE_NAME") != null
+            && !System.getenv("WEBSITE_SITE_NAME").isEmpty()) {
+            result = getAccessTokenOnAppService(resource);
+        } else {
+            result = getAccessTokenOnOthers(resource);
+        }
+        return result;
+    }
+
+    /**
+     * Get an access token.
+     *
+     * @param resource the resource.
+     * @param tenantId the tenant ID.
+     * @param clientId the client ID.
+     * @param clientSecret the client secret.
+     * @return the authorization token.
+     */
+    public String getAccessToken(String resource, String tenantId,
+                                 String clientId, String clientSecret) {
+        LOGGER.entering("AuthClient", "getAccessToken", new Object[] {
+            resource, tenantId, clientId, clientSecret });
+        LOGGER.info("Getting access token using client ID / client secret");
+        String result = null;
+
+        StringBuilder oauth2Url = new StringBuilder();
+        oauth2Url.append(OAUTH2_TOKEN_BASE_URL)
+                 .append(tenantId)
+                 .append(OAUTH2_TOKEN_POSTFIX);
+
+        StringBuilder requestBody = new StringBuilder();
+        requestBody.append(GRANT_TYPE_FRAGMENT)
+                   .append(CLIENT_ID_FRAGMENT).append(clientId)
+                   .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
+                   .append(RESOURCE_FRAGMENT).append(resource);
+
+        String body = post(oauth2Url.toString(), requestBody.toString(), "application/x-www-form-urlencoded");
+        if (body != null) {
+            JsonConverter converter = JsonConverterFactory.createJsonConverter();
+            OAuthToken token = (OAuthToken) converter.fromJson(body, OAuthToken.class);
+            result = token.getAccess_token();
+        }
+        LOGGER.log(FINER, "Access token: {0}", result);
+        return result;
+    }
+
+    /**
+     * Get the access token on Azure App Service.
+     *
+     * @param resource the resource.
+     * @return the authorization token.
+     */
+    private String getAccessTokenOnAppService(String resource) {
+        LOGGER.entering("AuthClient", "getAccessTokenOnAppService", resource);
+        LOGGER.info("Getting access token using managed identity based on MSI_SECRET");
+        String result = null;
+
+        StringBuilder url = new StringBuilder();
+        url.append(System.getenv("MSI_ENDPOINT"))
+           .append("?api-version=2017-09-01")
+           .append(RESOURCE_FRAGMENT).append(resource);
```

Comment:
I believe the resource string will need to be url encoded no?

Reply:
The KeyVaultClient code passes in a url encoded version so nothing further needs to be done